### ISSUE TITLE
isNodeJS check

### ISSUE THREAD

**Issue by dominataa** [kriskowal/q#645]
Hi, in asap.js you are checking whether the environment `isNodeJS`.
```
if (typeof process !== "undefined" && process.nextTick) {
    // Node.js before 0.9. Note that some fake-Node environments, like the
    // Mocha test runner, introduce a `process` global without a `nextTick`.
    isNodeJS = true;
}
```
We do have a fake environment with browserify where process.nextTick is used and this check is causing errors to be thrown synchronously causing issues.

One suggestion from ksmeltzer was to use `process.title === 'node'`. The problem is we have window in both envs and process etc..

**Comment by kitcambridge** [comment]
Or `Object.prototype.toString.call(process) == '[object process]'`...

**Comment by kriskowal** [comment]
Do these environments also provide a bad process.nextTick?

**Comment by kriskowal** [comment]
Come back and re-open this if process.nextTick is bad on these fake Node environments. The isNode test just means that Q will use process.nextTick and check whether process.domain exists when necessary. It should not have any harmful side effects.

**Comment by dominataa** [comment]
The problem is this in flush: 

``` javascript
           try {
                task();

            } catch (e) {
                if (isNodeJS) {
                    // In node, uncaught exceptions are considered fatal errors.
                    // Re-throw them synchronously to interrupt flushing!

                    // Ensure continuation if the uncaught exception is suppressed
                    // listening "uncaughtException" events (as domains does).
                    // Continue in next event to avoid tick recursion.
                    if (domain) {
                        domain.exit();
                    }
                    setTimeout(flush, 0);
                    if (domain) {
                        domain.enter();
                    }

                    throw e;

                } else {
                    // In browsers, uncaught exceptions are not fatal.
                    // Re-throw them asynchronously to avoid slow-downs.
                    setTimeout(function() {
                       throw e;
                    }, 0);
                }
            }
```

It thinks that the browser `isNodeJS` and is therefore throwing errors synchronously instead of async as it would otherwise and flush is not completing.

**Comment by slorber** [comment]
I modified Q locally and forced the use of:

```javascript
        requestTick = function () {
            setTimeout(flush, 0);
        };
```

and it works fine now! Q is not broken anymore.

So I think this is an issue that only happens when used with Browserify

**Comment by slorber** [comment]
This problem appeared between releases Browserify 8.0.3 and 8.1.0. Q works fine with older versions of Browserify

**Comment by slorber** [comment]
I've made a PR for node-process/browserify that solves this problem.
See https://github.com/defunctzombie/node-process/pull/37

**Comment by calvinmetcalf** [comment]
can confirm Q does some funky stuff in browserify because it thinks it's in node

**Comment by pho3nixf1re** [comment]
This happens with Webpack as well. It's killing our test suite and causing early exits instead of just test failures.

**Comment by kriskowal** [comment]
This has been addressed by @kahnjw in v1.2.1. Thanks for the tips.
